### PR TITLE
GEOT-6299 list functions

### PR DESCRIPTION
**Pull request by NielsCharlier**

### CODE REVIEW

**Comment by ianturton** [comment]
You need to update the user/artifacts/function_list - I think running mvn compile in docs/src/main should do it.

**Comment by NielsCharlier** [comment]
It's working now.

File: modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    @Override
+    public Object evaluate(Object feature) {
+        if (!(getExpression(0) instanceof PropertyName)) {
+            throw new IllegalArgumentException(
+                    "literate function first argument must be index property name");
+        }
+        PropertyName indexName = (PropertyName) getExpression(0);
+        Integer size = getExpression(1).evaluate(feature, Integer.class);
+        if (size == null) {
+            throw new IllegalArgumentException("literate function requires non-null size");
```
Comment (jnh5y) [inline modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java]: Does size need to be non-negative as well?

File: modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.LiteralExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.geotools.filter.visitor.DuplicatingFilterVisitor;
+import org.opengis.filter.capability.FunctionName;
+import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.ExpressionVisitor;
+import org.opengis.filter.expression.PropertyName;
+
+public class FilterFunction_literate extends FunctionExpressionImpl {
```
Comment (jnh5y) [inline modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java]: @NielsCharlier can you add a comment about what this function does?  From the name, I'm not sure.

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_litem.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.List;
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+public class FilterFunction_litem extends FunctionExpressionImpl {
+
+    public static FunctionName NAME =
+            new FunctionNameImpl(
+                    "litem",
```
Comment (jnh5y) [inline modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_litem.java]: While brevity is the soul of wit, I wonder if this is a bit too terse.  I'd suggest that all these functions are non-obvious and could use some documentation.

**Comment by NielsCharlier** [comment]
@jnh5y thank you for your review. You are of course right, and I have added the appropriate javadoc.

**Comment by NielsCharlier** [comment]
is this okay to be merged now?

File: modules/library/main/src/test/java/org/geotools/filter/function/ListFunctionsTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        assertTrue(value instanceof Integer);
+        assertEquals(4, ((Integer) value).intValue());
+    }
+
+    @Test
+    public void testLitem() {
+
+        List<Integer> list = Lists.newArrayList(1, 2, 3, 4);
+
+        Function exp = ff.function("litem", ff.property("."), ff.literal(2));
```
Comment (jnh5y) [inline modules/library/main/src/test/java/org/geotools/filter/function/ListFunctionsTest.java]: Thanks for pinging me again about this PR.

It looks like function starts counting at 0.  Is there a good way to document that?   

I can see starting either at 0 or 1.  As a programmer, 0 makes sense for arrays.  For lists, I find starting at 1 more natural.  

I'm completely fine either way; just looking for a way to communicate it to users.

**Comment by jnh5y** [comment]
> is this okay to be merged now?

I have one more question, but it isn't necessarily blocking.  I'd say this is ready to be merged.

**Comment by jnh5y** [comment]
(By the community rules, I don't know if I am permitted to merge this or not.)

**Comment by aaime** [comment]
By community rules I believe the module maintainer should merge. For this particular module the original maintainer is no longer available, so let's see who's been active in the last two years, as sort of a replacement for a maintainer:

```
> geotools/modules/extension/complex/src/main (master) $ git shortlog -nsu --since "two years ago" .
     5  Jim Hughes
     4  Andrea Aime
     3  Jody Garnett
     3  Torben Barsballe
     2  Nuno Oliveira
     1  Andrea Antonello
     1  Brad Hards
     1  James Hughes
     1  Niels Charlier
```

Well... you've been the most active dev :-D

**Comment by jnh5y** [comment]
@aaime Hmm... I racked up those 5 commits at the Java 11 code sprint.  If that qualifies me to merge this in good faith, I'm fine with that.  I'll note that I don't feel qualified to be the official maintainer, but I'm happy to 'help'.;)

While re-reading the comments, I noticed that Ian's suggestion about updating the document in docs/user/artifacts/function_list hasn't happened yet.  Once that happens, I'd be happy to merge.

**Comment by NielsCharlier** [comment]
> You need to update the user/artifacts/function_list - I think running mvn compile in docs/src/main should do it.

@ianturton I can't seem to figure this out. Building docs doesn't work. Inside src/main I can't call maven. Should I just do this manually or is it important to call something automised?

**Comment by NielsCharlier** [comment]
Nevermind, I found it. There is actually currently a PR to document how to change it: https://github.com/geotools/geotools/pull/2216

**Comment by NielsCharlier** [comment]
@jnh5y fixed

File: modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java (hunk 2 above)
Comment (jodygarnett) [reply modules/extension/complex/src/main/java/org/geotools/data/complex/expression/FilterFunction_literate.java]: I hate this class name, the originals are all made this way because they were generated. 

Please rename to LiterateFunction

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_size.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import org.geotools.filter.FunctionExpressionImpl;
+import org.geotools.filter.capability.FunctionNameImpl;
+import org.opengis.filter.capability.FunctionName;
+
+/**
+ * Return the size of a list (or other collection).
+ *
+ * @author Niels Charlier
+ */
+public class FilterFunction_size extends FunctionExpressionImpl {
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_size.java]: Please rename to SizeFunction

File: modules/library/main/src/test/java/org/geotools/filter/function/ListFunctionsTest.java (hunk 4 above)
Comment (jodygarnett) [reply modules/library/main/src/test/java/org/geotools/filter/function/ListFunctionsTest.java]: Document it in the function NAME descriptor

File: modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_litem.java (hunk 3 above)
Comment (jodygarnett) [reply modules/library/main/src/main/java/org/geotools/filter/function/FilterFunction_litem.java]: Yep, if we fill in the parameter descriptor information it can generate out nice information for WPS Capabilities and GeoTools function list.

File: docs/user/artifacts/function_list [hunk 6]
```diff
@@ -1014,6 +1014,15 @@ isometric(geometry, extrusion): isometric
 
 * isometric: Geometry 1:1 required
 
+lapply(source, expression): result
```
Comment (jodygarnett) [inline docs/user/artifacts/function_list]: These are generated from the code example here: 

https://github.com/geotools/geotools/blob/master/docs/src/main/java/org/geotools/main/FilterExamples.java#L527

**Comment by NielsCharlier** [comment]
changes applied as requested, thanks @jnh5y and @jodygarnett

File: docs/user/artifacts/function_list [hunk 7]
```diff
@@ -1066,6 +1075,26 @@ listMultiply(factor, list): listMultiply
...
+''''''''''''''''''''''''''''
+
+* source: List 1:1 required
+
+* index: Integer 1:1 required
+
+* result
+
+literate(index, times, expression): result
+''''''''''''''''''''''''''''''''''''''''''
```
Comment (jodygarnett) [inline docs/user/artifacts/function_list]: The description from the NAME should show up here? `List of items resulted from iteration.`

We may need to rerun the generator from https://github.com/geotools/geotools/pull/2216

**Comment by jodygarnett** [comment]
This looks good on my end, ... 

I would like to know about a) build box and b) deprecation api removal branch before merging anything on this codebase.

**Comment by NielsCharlier** [comment]
@jodygarnett thanks. so I wait until api removal branch is merged?

**Comment by aaime** [comment]
The PR is referring to a geoserver issue about CSW, resulting in:
* Confusion looking at the GIT history, since this is Geotools
* Improper reporting in release notes, since the change will show up only in GeoServer

Can you please:
* Create a ticket in GeoTools too
* When you merge, do a "squash and merge" and change the title so that it refers to the geotools ticket?

@jodygarnett about the deprecation, I don't mind if this gets merged and causes minor conflicts, I'll sort them out, as indicated on list I need to merge on a weekend but won't block the entire project for it
However, we need the build servers online, so that we can notice eventual downstream issues right away, instead of having changes pile up for days and then having to figure out what broke GWC or GeoServer.

Long story short, as far as I'm concerned, feel free to merge as soon as build.geoserver.org is back online, and properly working.

**Comment by NielsCharlier** [comment]
Thanks to everyone for their reviews
